File: Day #1091/solution.go
// Count connected components via Union-Find with path compression. Time ~O(N+E*alpha), Space O(N).
package main

import "fmt"

var parent map[int]int

func find(x int) int {
	for parent[x] != x {
		parent[x] = parent[parent[x]]
		x = parent[x]
	}
	return x
}

func unite(a, b int) { parent[find(a)] = find(b) }

func countGroups(adj map[int][]int) int {
	parent = make(map[int]int)
	for u := range adj {
		parent[u] = u
	}
	for u, nbrs := range adj {
		for _, v := range nbrs {
			unite(u, v)
		}
	}
	roots := map[int]bool{}
	for u := range adj {
		roots[find(u)] = true
	}
	return len(roots)
}

func main() {
	adj := map[int][]int{
		0: {1, 2}, 1: {0, 5}, 2: {0}, 3: {6}, 4: {}, 5: {1}, 6: {3}}
	fmt.Println(countGroups(adj))
}

File: Day #1091/solution.js
// Count connected components via Union-Find with path compression. Time ~O(N+E*alpha), Space O(N).
function countGroups(adj) {
  const parent = new Map();
  for (const u of adj.keys()) parent.set(u, u);
  function find(x) {
    while (parent.get(x) !== x) {
      parent.set(x, parent.get(parent.get(x)));
      x = parent.get(x);
    }
    return x;
  }
  function unite(a, b) { parent.set(find(a), find(b)); }
  for (const [u, nbrs] of adj) for (const v of nbrs) unite(u, v);
  const roots = new Set();
  for (const u of adj.keys()) roots.add(find(u));
  return roots.size;
}

const adj = new Map([
  [0, [1, 2]], [1, [0, 5]], [2, [0]], [3, [6]], [4, []], [5, [1]], [6, [3]]]);
console.log(countGroups(adj));
